improve target coordinate specification in jplhorizons

m-stclair opened the pull request:
closes https://github.com/astropy/astroquery/issues/1555
Fixes #2599.

m-stclair commented:
hi @bsipocz -- I've reimplemented these features, added basic test coverage for them, updated other tests as necessary, and updated docstrings. I'm assuming the fast CI failures where the pipeline can't find `tox` are because this is a draft; if not, I have no idea how to interrogate those. What else do you think needs to be done before I mark this as ready for review?

bsipocz commented:
It's a weird and new failure about tox, the draft status of the PR shouldn't affect the CI this way. I'll look into it and fix it separately.

If the content is ready, then do mark it ready for review, and we can ask @mkelley to do a first round of review while I'm dealing with CI.
(Tests can be checked locally, including the remote tests, etc.)

bsipocz commented:
@m-stclair - could you rebase? CI has been fixed, but I cannot step in as a maintainer to do the rebase as the PR was opened from an org fork rather than from yours.

m-stclair commented:
Ty for fixing the tox issue. I have rebased and am now marking this ready for review.

bsipocz commented on astroquery/jplhorizons/tests/test_jplhorizons.py:
```diff
@@ -256,22 +256,23 @@ def test_vectors_query_payload():
                                    get_query_payload=True)
-
     assert res == OrderedDict([
-        ('format', 'text'),
```
Please refrain from reformatting like this. Add the new dict element, but leave the rest alone.

m-stclair replied:
Switching the order back will cause the test to fail, because it is an equality test against an `OrderedDict`, and changes I need to make in order to construct the payload to include target coordinates alter the order in which keys are inserted in the `OrderedDict`.

m-stclair replied:
will add a changelog entry and update narrative docs tomorrow or tuesday.

bsipocz replied:
I don't mean the order, but don't do reformatting, no need to change 15 lines when only one or two need to be shuffled around

m-stclair replied:
oh you mean the indentation. sorry, didn't even notice.

bsipocz commented on astroquery/jplhorizons/core.py:
```diff
@@ -1181,14 +1224,23 @@ def _parse_result(self, response, verbose=None):
+                else:
+                    headerline[3] = 'interfering_body'
                 headerline[-1] = '_dump'
+                if (
+                    isinstance(self.id, dict) or str(self.id).startswith('g:')
+                ):
```
We have longer than 80 char lines, do keep this in one line instead of 1 with information and 2 of fillers. This is for here and a few other places, too.

m-stclair replied:
copy that.

bsipocz commented on astroquery/jplhorizons/core.py:
```diff
@@ -1060,18 +1075,17 @@ def vectors_async(self, *, get_query_payload=False,
-        if isinstance(self.location, dict):
-            raise ValueError(('cannot use topographic position in state'
-                              'vectors query'))
+        #
+        # if isinstance(self.location, dict):
+        #     raise ValueError(('cannot use topographic position in state'
+        #                       'vectors query'))
```
Do remove unused code rather than comment out

m-stclair replied:
oops. missed this.

bsipocz commented on astroquery/jplhorizons/core.py:
```diff
@@ -580,19 +601,9 @@ def ephemerides_async(self, *, airmass_lessthan=99,
-            request_payload['SITE_COORD'] = "'{:f},{:f},{:f}'".format(
-                self.location['lon'], self.location['lat'],
-                self.location['elevation'])
+            request_payload = dict(
+                **request_payload, **self._location_to_params(self.location)
+            )
```
keep it in one line

m-stclair replied:
copy that.

bsipocz commented on astroquery/jplhorizons/core.py:
```diff
@@ -15,6 +12,9 @@
 from astropy.time import Time
 from astropy.utils.exceptions import AstropyDeprecationWarning
 from astropy.utils.decorators import deprecated_renamed_argument, deprecated_attribute
+from numpy import nan
```
if you need to reorder imports, keep the numpy ones before astropy.

m-stclair replied:
copy that.

bsipocz commented on astroquery/jplhorizons/core.py:
```diff
@@ -1032,17 +1043,21 @@ def vectors_async(self, *, get_query_payload=False,
         if self.id is None:
             raise ValueError("'id' parameter not set. Query aborted.")
+        elif isinstance(self.id, dict):
+            commandline = (
+                f"g:{self.id['lon']},{self.id['lat']},"
+                f"{self.id['elevation']}@{self.id['body']}"
+            )
```
this also probably fit the line lenght limit

m-stclair replied:
condensed to one line.

bsipocz reviewed (changes requested):
Some minor nitpicks, and please add a changelog entry.  

Also, please update the narrative docs, changes here have effect on some of the doctest expectations there.

m-stclair commented:
will add a changelog entry and update narrative docs tomorrow or tuesday.

bsipocz commented:
> The codecov misses are pretty small things, but let me know if it's important to cover them and I'll add a couple of little unit tests!

codecov is red herring here, it doesn't look for the changes with the remote tests, so it misses the actual ~2% test increase

m-stclair commented:
on codecov: got it, makes sense

m-stclair commented:
@bsipocz , I believe I've addressed all your requests.

eerovaher commented on astroquery/jplhorizons/core.py:
```diff
@@ -1032,17 +1041,18 @@ def vectors_async(self, *, get_query_payload=False,
         if self.id is None:
             raise ValueError("'id' parameter not set. Query aborted.")
+        elif isinstance(self.id, dict):
+            commandline = f"g:{self.id['lon']},{self.id['lat']},{self.id['elevation']}@{self.id['body']}"
```
Usually I prefer f-strings over `str.format()`, but in this case the following can be considered:
```suggestion
            commandline = "g:{lon},{lat},{elevation}@{body}".format(**self.id)
```

m-stclair replied:
this is prettier, yes.

eerovaher commented on astroquery/jplhorizons/core.py:
```diff
@@ -1132,6 +1141,30 @@ def vectors_async(self, *, get_query_payload=False,
+    @staticmethod
+    def _prep_loc_dict(loc_dict, attr_name):
+        """prepare coord specification dict for 'location' or 'id'"""
+        if {'lat', 'lon', 'elevation'} - set(loc_dict.keys()) != set():
```
An empty set is falsy and a non-empty set is truthy. And the explicit conversion of dict keys to `set` is not needed (at least in Python 3.10):
```suggestion
        if {'lat', 'lon', 'elevation'} - loc_dict.keys():
```

m-stclair replied:
Huh, I didn't know that about `dict_keys`, neat -- set operators in general fail with non-set iterables and I'd never tried. 

I have an allergy to truthiness tests but I understand that people often prefer them. Is your preference here strong?

m-stclair replied:
Actually, my preference isn't _that_ strong. I'll just change this.

eerovaher commented on astroquery/jplhorizons/core.py:
```diff
@@ -1181,14 +1214,21 @@ def _parse_result(self, response, verbose=None):
                 headerline = str(line).split(',')
                 headerline[2] = 'solar_presence'
-                headerline[3] = 'flags'
+                if 'Earth' in centername:
+                    headerline[3] = 'lunar_presence'
+                else:
+                    headerline[3] = 'interfering_body'
```
Is the one-liner too long?
```suggestion
                    headerline[3] = "lunar_presence" if "Earth" in centername else "interfering_body"
```

m-stclair replied:
it was too long for what I thought the line length limit was when I was writing it -- most of the module had been wrapped at 79 -- but I have since learned that this is not a real `astroquery` restriction and that longer lines are in fact preferred. So this would work.

eerovaher commented on astroquery/jplhorizons/tests/test_jplhorizons_remote.py:
```diff
@@ -405,3 +409,51 @@ def test_ephemerides_extraprecision(self):
+        """
+        phobos = {
+            'body': 401,
+            'lon': random.randint(-150, -1),
+            'lat': random.randint(-80, 80),
```
I'm not sure running the tests on random inputs is worth the effort. Even if the tests stumble upon a weird edge case that causes the tests to fail, it can be very difficult to figure out what the problematic values were, which prevents them from being addressed.

m-stclair replied:
It makes me nervous to use only white-box tests, but I understand if you prefer it for stability reasons. So I will turn this into another explicit equality test.

eerovaher replied:
The best way of doing randomized tests would be to use [`hypothesis`](https://hypothesis.readthedocs.io/en/latest/) because it is capable of storing inputs that caused a failure, but `astroquery` isn't using it at the moment.

m-stclair replied:
I do like `hypothesis`. But this is totally fine with me. Weird edge cases are not actually that likely here.

eerovaher commented on docs/jplhorizons/jplhorizons.rst:
```diff
@@ -72,6 +105,24 @@ as the observer's location:
+    >>> obj = Horizons(id=double, location=ce_2, epochs=2454483.84247)
+    >>> vecs = obj.vectors()
+    >>> distance_km = (vecs['x'] ** 2 + vecs['y'] ** 2 + vecs['z'] ** 2) ** 0.5 * 1.496e8
+    >>> print(round(distance_km.value.data[0], 3))
```
```suggestion
    >>> print(f"{distance_km.value.data[0]:.3f}")
```

m-stclair replied:
better, yes

m-stclair commented:
@eerovaher , I believe I've addressed all your suggestions. I do not understand why the readthedocs build is now failing, but it does not appear to be related to any actual code changes.

eerovaher commented:
I don't really understand why the documentation build failed. Perhaps if you squash some of the commits (not necessarily down to one) and then force-push to GitHub the next documentation build might succeed.

bsipocz commented:
> I do not understand why the readthedocs build is now failing, but it does not appear to be related to any actual code changes.

This is a weird problem, I don't think I saw it before (the commit hashes match, yet RTD complains that it cannot check it out). A restart didn't resolve it, so I would say wait it out while I investigate whether other repos are affected, too, or try with the squashing idea.

bsipocz commented:
OK, leave the RTD for now, this is a more widespread problem and not much we can do about it in this PR.

earthshrink commented:
hi @m-stclair, in first comment above, did you mean closes #2599?

m-stclair commented:
No, it's just the confusing GitHub convention where PRs and issues are notated the same way, and this is PR 2625. When I opened this PR I hadn't noticed you'd opened that issue. However, it's definitely true that this closes #2599.

bsipocz commented:
Sorry, that was my bad, I indeed meant to add https://github.com/astropy/astroquery/issues/2599 there and not the PR's number. Thanks for spotting it. 😄

m-stclair commented:
Nevermind then! I thought it had been automatically added.

On Thu, Dec 15, 2022 at 5:46 PM Brigitta Sipőcz ***@***.***>
wrote:

> Sorry, that was my bad, I indeed meant to add #2599
> <https://github.com/astropy/astroquery/issues/2599> there and not the
> PR's number. Thanks for spotting it. 😄
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/astropy/astroquery/pull/2625#issuecomment-1353478245>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APIXBZIVPJAPTAGVI3XA5MTWNNKOJANCNFSM6AAAAAAS2QP3O4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

bsipocz commented:
Yes, your original PR predates all the new issues. And based on that, this would also address https://github.com/astropy/astroquery/issues/1555

bsipocz commented:
Thanks @m-stclair!
